src/index.js

```diff
@@ -6,229 +6,125 @@
 const $classConstructor = '__classConstructor';
 const $classMethod = '__classMethod';
 const $objectMethod = '__objectMethod';
+const $bindedArrowFunction = '__bindedArrowFunction';
 
-export default function build (babel: Object): Object {
-  const {types: t, traverse} = babel;
+export default function build(babel:Object):Object {
+  const {types: t} = babel;
 
-  const referenceVisitor = {
-    enter (path, state) {
-      const node = path.node,
-        scope = path.scope;
-      if (!path.isJSXIdentifier() && path.isIdentifier()) {
-        // direct references that we need to track to hoist this to the highest scope we can
-        if (path.isReferenced() || path.parent.type == 'AssignmentExpression') {
-          const bindingInfo = scope.getBinding(node.name);
-
-          // this binding isn't accessible from the parent scope so we can safely ignore it
-          // eg. it's in a closure etc
-          if (bindingInfo !== state.scope.getBinding(node.name)) {
+  return {
+    visitor: {
+      Function: {
+        exit (path) {
+          const {node} = path;
+          if (node[$classConstructor] || node.body[$classConstructor] || node[$classMethod] || node[$objectMethod] || node[$bindedArrowFunction]) {
             return;
           }
-
-          if (bindingInfo && bindingInfo.constant) {
-            state.bindings[node.name] = bindingInfo;
+          if (path.findParent(({node}) => node._generated || node._compact)) {
```

Comment:
Not actually a bad idea, i do that with the typecheck plugin - run it against itself as an extra sanity check.